client/src/main/java/io/pravega/client/stream/impl/StreamCut.java

```diff
@@ -29,5 +32,15 @@
     private final Stream stream;
     @Getter(value = AccessLevel.PACKAGE)
     private final Map<Segment, Long> positions;
-    
+
```

Comment:
we should add another validation to make sure we dont have overlapping segments in a stream cut.

Reply:
I suggest we make `StreamCut` implement `Comparable` so that we can compare two cuts. Is that ok? If so, I can make the change.

Reply:
Actually, comparable is not going to do it because the set of stream cuts is a partial order. When given two cuts C1 and C2, they can be:

- Equal
- C1 > C2: for each segment of C1, either it succeeds some segment of C2 or it is in C2 with a smaller offset.
- C1 < C2: for each segment of C2, either it succeeds some segment of C1 or it is in C1 with a smaller offset.
- C1 and C2 are incomparable

To be able to compare cuts, I need four states, so I can't simply use negative, zero, positive like with the comparable spec.

Reply:
Actually the successor and predecessor information is not present with the client directly and controller can and will do that bit of validation.

What i was suggesting was to take all segments in the cut and make sure none of them overlap. 
And make sure there are no gaps either.
For example, say a cut is made up of S1 S2, S3... Sn.
We should make sure Union(S1.keyrange, S2.keyrange, S3.keyrange, ... Sn.keyrange) = (0.0, 1.0) and Intersection(S1.keyrange, S2.keyrange, S3.keyrange, ... Sn.keyrange) = null

Reply:
The invariant you are suggesting is important even for other things we do like checkpoints.

I don't have range information readily available. `Segment` doesn't contain it, and I don't think a cut should be more than a list of segments and corresponding offsets, so without adding range information to the cut, how can I obtain the corresponding ranges?

Reply:
its true that Segment object being passed here does not have the information. Though this information is available with clients (readers and writers) but those clients may have moved ahead from these segments being used in the cut and have forgotten about their ranges. 

And you are right about not putting this information in streamCut. Just segment-id/offset suffices. 

So far as validating a stream cut for truncation, we are doing this validation on Controller. 
However, as you said, checkpoint should also verify this invariant. @tkaitchuck @shrids any ideas?

Reply:
one simple way would be to have a call in the controller to get the range of a segment, like `getSegmentRange(Segment segment)`. If we have that, I can check the invariant in the test case.

Reply:
Create this issue for it:

https://github.com/pravega/pravega/issues/2023